### PR TITLE
Core: Replace preset-env polyfills with babel-polyfills

### PR DESCRIPTION
**Pull request by 43081j**
Issue: #11255.

Another attempt, this time it depends on babel/babel-polyfills#40

presumably what they will want us to do is use the babel-polyfills plugin rather than the preset-env corejs config. the docs seem to suggest thats the way forward, so this is just a draft of doing it that way

### CODE REVIEW

**Comment by nicolo-ribaudo** [comment]
> presumably what they will want us to do is use the babel-polyfills plugin rather than the preset-env corejs config

Yes, this will be the recommended way starting from Babel 8 :+1:

I want to add that even if this package's version is currently `0.0.6`, it passes all the `core-js` tests we have in `@babel/preset-env`. The only reason for a `0.0.x` version is that we'll release it as stable and officially make everyone migrate only once we release Babel 8.

The way you migrate to the new plugin in this PR is correct.

**Comment by 43081j** [comment]
awesome, thanks for taking time to explain that.

and from what i can gather, setting the new flag to `true` means the paths are resolved from the cwd?

**Comment by nicolo-ribaudo** [comment]
> and from what i can gather, setting the new flag to true means the paths are resolved from the cwd?

It is relative to the config file where storybook's preset is enabled, or to cwd if the preset is enabled in the programmatic options.

I don't know how you enable this preset, but maybe it would be better to:
- Add `core-js` as a dependency to Storybook
- Use `absoluteImports: require.resolve("core-js")`, so that it always resolves Storybook's `core-js` dependency.

**Comment by 43081j** [comment]
yeah thats a good idea i think.

we already have core-js as a dependency but should probably resolve to it here rather than leave it up to babel. ill change it

**Comment by merceyz** [comment]
Should be fine to remove `core-js` as a dependency in the e2e tests now
ref https://github.com/storybookjs/storybook/blob/6a3bf803337c877ca83e17784e7a73fe82240d74/scripts/run-e2e-config.ts#L168-L169

**Comment by 43081j** [comment]
done.

also i noticed almost every package we have has a core-js dependency. are they still needed?

**Comment by 43081j** [comment]
@nicolo-ribaudo do we have any timeline for when the merged pr in babel-polyfills will be published? just so i know to keep an eye on it if its soon. or if you could let me know in here when it is, that'd be super useful

**Comment by nicolo-ribaudo** [comment]
I'm updating some deps and then I'll release.

**Comment by nicolo-ribaudo** [comment]
Released

**Comment by 43081j** [comment]
great, thanks for the quick turnaround on that :D 

have pushed the updated one here too

File: package.json
Comment (nicolo-ribaudo) [inline package.json:143]: This needs to be moved to https://github.com/storybookjs/storybook/blob/next/lib/core/package.json

Comment (43081j) [reply package.json:143]: good point, will move it

**Comment by 43081j** [comment]
i have tried this in two test repos:

* one where `core-js@2` is a top level dependency, currently broken with storybook master
* one where `core-js@3` is a top level dependency, currently works with master

with this branch, both worked fine! maybe we have fixed it finally? 😂

File: scripts/run-e2e-config.ts
Comment (merceyz) [inline scripts/run-e2e-config.ts:163]: I meant to remove all mentions of `core-js` from this file, with the change in this PR it Should Work™

**Comment by 43081j** [comment]
i do still need to know if all the sub-packages can drop the corejs dependency too.

pretty much every package we have has a corejs dependency in their package.json (every app/, addon/, etc).

**Comment by merceyz** [comment]
> also i noticed almost every package we have has a core-js dependency. are they still needed?

Depends if they actually use it or if it was a workaround to get it to work before this PR

**Comment by 43081j** [comment]
in the last commit i dropped it as a dependency from everything that had it except:

* those which reference it explicitly
* those which reference the corejs setting of babel

its in one commit in case we can't do that / it isn't so simple. could do with some help there, to confirm it won't just cause chaos

**Comment by merceyz** [comment]
I'd like to have the PnP test give this a go before its merged, it will catch any dependency issues that might exist. 
Working on fixing it in https://github.com/storybookjs/storybook/pull/13059

**Comment by gaetanmaisse** [comment]
> pretty much every package we have has a corejs dependency in their package.json (every app/, addon/, etc).

> Depends if they actually use it or if it was a workaround to get it to work before this PR

Almost every package is (was? with this PR) doing some `require("core-js...")` in the compiled files even if it's not imported in the sources files, for instance in `addons/a11y/dist/index.js`:
```js
require("core-js/modules/es.array.slice");
require("core-js/modules/es.object.define-properties");
require("core-js/modules/es.object.freeze");
```

Taking a look at the `dist` folder of some `addons` and `lib` should let you know if `core-js` is still needed as regular dep :)

**Comment by 43081j** [comment]
Yes that is exactly why i left those which reference corejs in their babel config still.

I only removed it as a dependency from the packages which definitely do not use it (directly or via babel).

**Comment by 43081j** [comment]
whats the best way forward with this now?

do we just need review? is there any solid reliable way of testing those package manifest changes?

**Comment by shilman** [comment]
@43081j we're in RC for 6.1 and this is a destabilizing change, so we probably can't merge it until that's released (hopefully this week). Is it a breaking change?

**Comment by 43081j** [comment]
its not necessarily a breaking change, but i suppose we have technically changed how things get polyfilled in the generated code. so probably a good idea to wait

**Comment by shilman** [comment]
@43081j `6.1` has been released, so that's no longer blocking this PR. we can potentially merge it and release it as `6.2-alpha`. Before we do that, though, I think we need to figure out what the impact of this PR is.

Strawman: merge the PR, release as `6.2-alpha`, revert/adapt if it breaks people. WDYT @ndelangen?

**Comment by ndelangen** [comment]
honestly not sure @shilman

**Comment by 43081j** [comment]
the manifest changes are the only part im unsure of, as it seems at some point we just added core-js to every package as a catch-all ("make it work" fix).

but tests won't really catch any issues with that as this repo will always have a resolvable core-js as a whole.

**Comment by merceyz** [comment]
> but tests won't really catch any issues with that as this repo will always have a resolvable core-js as a whole.

That's not really true, the PnP tests will not allow access to `core-js` if it isn't declared as a dependency. That only works for what it covers ofc which is vue and cra https://github.com/storybookjs/storybook/blob/f4535542e2136b629cb5649e6021de066dbaad2b/.circleci%2Fconfig.yml#L261

Since the PnP tests have been fixed in next, if you merge/rebase this PR we can see if it works or not

File: lib/core/src/server/common/babel.ts
Comment (merceyz) [inline lib/core/src/server/common/babel.ts:40]: @nicolo-ribaudo Is there a way to give it the version of `core-js` or does it figure that out itself? Not all versions of `core-js` have the same imports :thinking:

File: lib/core/src/server/common/babel.ts
Comment (merceyz) [inline lib/core/src/server/common/babel.ts:38]:
```suggestion
      absoluteImports: require.resolve('core-js'),
      version: require('core-js/package.json').version
```

Comment (43081j) [reply lib/core/src/server/common/babel.ts:38]: the plugin itself is a corejs 3 plugin, im not sure this is a thing anymore. it existed in the preset-env stuff because it tried to handle both at once, but this doesn't, we've explicitly imported the corejs3 plugin

Comment (merceyz) [reply lib/core/src/server/common/babel.ts:38]: Yes, but different versions of the 3.x release has different polyfills

Comment (nicolo-ribaudo) [reply lib/core/src/server/common/babel.ts:38]: It's still needed: this plugin outputs code that is compatible with every `core-js@3` version, so it cannot inject imports to polyfills added, for example, in `core-js@3.7` because that would break with `core-js@3.6`.

Comment (43081j) [reply lib/core/src/server/common/babel.ts:38]: it amazes me that core-js' minor versions are incompatible. i will add the change later today

Comment (merceyz) [reply lib/core/src/server/common/babel.ts:38]: Not incompatible, just that newer versions have more polyfills. Without specifying a version the plugin will assume 3.0 and not polyfill newer features

Comment (43081j) [reply lib/core/src/server/common/babel.ts:38]: true i guess the breaking change would be on babel's end, to start pulling in the new imports. which is why we have to tell it which version, since they'd have to track the latest otherwise and enforce it

Comment (merceyz) [reply lib/core/src/server/common/babel.ts:38]: > to start pulling in the new imports

That's why they don't, they only pull in what would be in version 3.0 if no version is specified

**Comment by 43081j** [comment]
have rebased and added the proposed corejs version config.

lets see what CI says this time 🤔

**Comment by 43081j** [comment]
all tests pass for me locally.

though i am seeing these warnings (im not sure if they existed before):

```
  When setting `useBuiltIns: 'usage'`, polyfills are automatically imported when needed.
  Please remove the direct import of `core-js` or use `useBuiltIns: 'entry'` instead.
```

The only place i can think of that is doing direct imports still is here:
https://github.com/storybookjs/storybook/blob/next/app/angular/src/client/preview/angular-polyfills.ts

can this go away? now that the core config will polyfill it?

**Comment by nicolo-ribaudo** [comment]
`reflect-metadata` won't be automatically polyfilled because it's just an early proposal.
If the warning wasn't there with `@babel/preset-env`, I can try to align the warning logic so that both do the same thing.

**Comment by ndelangen** [comment]
>  (im not sure if they existed before)

they did

**Comment by ndelangen** [comment]
I think this might fix the issue currently experienced in https://github.com/storybookjs/storybook/pull/13013

**Review by ndelangen — APPROVED** [review]

File: scripts/run-e2e-config.ts
Comment (merceyz) [inline scripts/run-e2e-config.ts:146]:
```suggestion
```
This has been fixed

**Comment by merceyz** [comment]
The packages that are transpiled before publish will need to declare `core-js` as a dependency still, so that probably needs to be reverted

**Comment by ndelangen** [comment]
@shilman let's merge?

**Review by merceyz — APPROVED** [review]
PnP is happy so LGTM :+1:

**Comment by shilman** [comment]
@ndelangen I'm going to make the cut over to 6.2 on Monday. Will merge then.

**Review by merceyz — CHANGES_REQUESTED** [review]
Need to remove `useBuiltIns: 'usage', corejs: '3'` from here as well https://github.com/storybookjs/storybook/blob/3239d259f1aa73cb556ded32df83f6039751b155/lib%2Fcore%2Fsrc%2Fserver%2Fcommon%2Fes6Transpiler.ts#L18

**Comment by ndelangen** [comment]
👏

**Pull request merged**
